### PR TITLE
Convert task_spec to flatbuffers

### PR DESCRIPTION
**Pull request by pcmoritz**
Note we compile in `DEBUG` mode in the valgrind tests because there seems to be an uninitialized value error when compiling in `RELEASE` mode.

### CODE REVIEW

File: src/common/task.h [hunk 1]
```diff
@@ -307,21 +271,13 @@ double task_spec_get_required_resource(const task_spec *spec,
 ObjectID task_compute_put_id(TaskID task_id, int64_t put_index);
 
 /**
- * Free a task_spec.
- *
- * @param The task_spec in question.
- * @return Void.
- */
-void free_task_spec(task_spec *spec);
```
Comment (robertnishihara) [inline src/common/task.h]: why get rid of `free_task_spec`?

File: src/local_scheduler/local_scheduler_algorithm.cc [hunk 2]
```diff
@@ -841,20 +866,23 @@ void handle_actor_task_submitted(LocalSchedulerState *state,
      * will be resubmitted (internally by the local scheduler) whenever a new
      * actor notification arrives. */
     utarray_push_back(algorithm_state->cached_submitted_actor_tasks, &spec);
+    utarray_push_back(algorithm_state->cached_submitted_actor_task_sizes,
```
Comment (robertnishihara) [inline src/local_scheduler/local_scheduler_algorithm.cc]: Why do we need to store the `task_spec` sizes separately from the task spec? Can we provide a method for computing it or even store it as a field in the task spec?

File: src/common/common_protocol.h [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+#include "common.h"
+
+/**
+ * Convert an object ID to a flatbuffer string.
+ *
+ * @param fbb Reference to the flatbuffer builder.
+ * @param object_id The object ID to be converted.
+ * @return The flatbuffer string contining the object ID.
+ */
+flatbuffers::Offset<flatbuffers::String> to_flat(
```
Comment (robertnishihara) [inline src/common/common_protocol.h]: I realize these methods were just moved around, but these names should probably be a bit more descriptive.

File: src/common/format/common.fbs [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+  function_id: string;
+  // Task arguments.
+  args: [Arg];
+  // Object IDs of return values.
+  returns: [string];
+  // Resource vector for this task. A resource vector maps a resource index
+  // (like "cpu" or "gpu") to the number of units of that resource required.
+  // Note that this will allow us to support arbitrary attributes:
+  // For example, we can have a coloring of nodes and "red" can correspond
+  // to 0.0, "green" to 1.0 and "yellow" to 2.0.
```
Comment (robertnishihara) [inline src/common/format/common.fbs]: The documentation for the required resources is confusing. Maybe something like

"The required_resources vector indicates the quantities of the different resources required by this task. The index in this vector corresponds to the resource type defined in the ResourceIndex enum. For example, required_resources[0] is the number of CPUs required, and required_resources[1] is the number of GPUs required".

**Review by robertnishihara — COMMENTED** [review]
Nice job!

File: src/local_scheduler/local_scheduler_algorithm.cc (hunk 2 above)
Comment (pcmoritz) [inline src/local_scheduler/local_scheduler_algorithm.cc]: We only have access to the size once the flatbuffer builder has been finalized, so we can't store it in the message itself.

**Pull request merged**